File: FontanaEsame/Implementazioni/Regressione/MatLab/Regressione.m
dati = readtable('winequality-red.csv', 'Delimiter', ';');

rng(5); %Random seed per riproducibilità
cv = cvpartition(size(dati, 1), 'Holdout', 0.2); %Split 80-20% tra training e test set

% Training set
X_train = dati(cv.training, 1:11); % Prendi le colonne da 1 a 11 per le features di input
y_train = dati(cv.training, 12); % Prendi la colonna 12 per la feature di output

% Testing set
X_test = dati(cv.test, 1:11); 
y_test = dati(cv.test, 12); 

nd = size(X_train, 1); % numero dati di training

% Costruzione della matrice dei predittori per i dati di addestramento
A = ones(nd, 12); % Creazione di una matrice di tutti 1 con 12 colonne
%nelle coolonne da 2 a 12 inseriscoo i dati di X_train
A(:, 2:12) = table2array(X_train);
yest = table2array(y_train); % Dati noti

%Risoluzione del problema ai minimi quadrati (calcolo la QR economica)
tic;
[Q,R]=qr(A,'econ');

soluzione = R \ (Q' * yest);
tempo = toc;
fprintf('I coefficienti della regressione con QR sono:\n');
disp(soluzione);
fprintf('Il tempo impiegato per la risoluzione con QR è: %f\n', tempo);


%risoluzione del problema ai minimi quadrati con la SVD
tic;
[U, S, V] = svd(A, 0);
soluzione = V * (S \ (U' * yest));
tempo = toc;
fprintf('I coefficienti della regressione con SVD sono:\n');
disp(soluzione);
fprintf('Il tempo impiegato per la risoluzione con SVD è: %f\n', tempo);


%risoluzione con equazioni normali
tic;
soluzione = (A' * A) \ (A' * yest);
tempo = toc;
fprintf('I coefficienti della regressione con equazioni normali sono:\n');
disp(soluzione);
fprintf('Il tempo impiegato per la risoluzione con equazioni normali è: %f\n', tempo);


%TESTING -> predico i valori di y per i dati di test, li confronto con i valori noti e calcolo l'errore RMSE (Root Mean Squared Error)
nd = size(X_test, 1); % numero dati di test
A = ones(nd, 12); % Creazione di una matrice di tutti 1 con 12 colonne
A(:, 2:12) = table2array(X_test); % inserisco i dati di X_test

yest = A * soluzione; % predizione dei valori di y
y_test = table2array(y_test); % valori noti

errore = sqrt(mean((yest - y_test).^2)); % calcolo dell'errore RMSE
fprintf("L'errore RMSE è: %f\n", errore);

File: FontanaEsame/Implementazioni/Regressione/Python/Regressione.py
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import time

data = pd.read_csv('winequality-red.csv', delimiter=';')


X = data.iloc[:, :-1]
y = data.iloc[:, -1]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=5)

#Ai fini della regressione lineare, aggiungiamo una colonna di 1 per il termine noto
X_train = np.hstack((np.ones((X_train.shape[0], 1)), X_train.values))
X_test = np.hstack((np.ones((X_test.shape[0], 1)), X_test.values))

# QR
start_time = time.time()
Q, R = np.linalg.qr(X_train)
sol_qr = np.linalg.solve(R, Q.T @ y_train)
qr_time = time.time() - start_time
print("Soluzione con QR:")
print(sol_qr)
print(f'Tempo per QR: {qr_time:.6f} secondi')

# SVD
start_time = time.time()
U, S, VT = np.linalg.svd(X_train, full_matrices=False)
sol_svd = VT.T @ np.linalg.solve(np.diag(S), U.T @ y_train)
svd_time = time.time() - start_time
print('Soluzione con SVD:')
print(sol_svd)
print(f'Tempo per SVD: {svd_time:.6f} secondi')

# Equazioni normali
start_time = time.time()
sol_ne = np.linalg.solve(X_train.T @ X_train, X_train.T @ y_train)
ne_time = time.time() - start_time
print('Soluzione con equazioni normali:')
print(sol_ne)
print(f'Tempo per equazioni normali: {ne_time:.6f} secondi')

# RMSE
y_pred = X_test @ sol_ne
rmse = np.sqrt(np.mean((y_pred - y_test) ** 2))
print(f'RMSE: {rmse:.6f}')
